syntax = "proto3";

package fp_network;

message ClientStreamState {
    enum State {
        READY_FOR_PPS_SPS_IDR = 0;
        READY_FOR_VIDEO = 1;
    }
    State state = 1;
    uint32 stream_num = 2;
}

message ClientState {
    enum State {
        DISCONNECTING = 0;
    }
    State state = 1;
}

message KeyboardFrame {
    uint32 key = 1;
    bool pressed = 2;
}

message MouseFrame {
    uint32 mouse_x = 1;
    uint32 mouse_y = 2;
    uint32 stream_num = 3;
    enum MouseButtons {
        MOUSE_L = 0;
        MOUSE_MIDDLE = 1;
        MOUSE_R  = 2;
        MOUSE_X1 = 3;
        MOUSE_X2 = 4;
    }
    optional MouseButtons button = 4;
    optional bool pressed = 5;
    optional double mouse_wheel_x = 6;
    optional double mouse_wheel_y = 7;
}

message ControllerFrame {
    uint32 sequence_num = 1;
    uint32 w_buttons = 2;
    uint32 b_left_trigger = 3;
    uint32 b_right_trigger = 4;
    uint32 s_thumb_LX = 5;
    uint32 s_thumb_LY = 6;
    uint32 s_thumb_RX = 7;
    uint32 s_thumb_RY = 8;
}

message RequestToHost {
    enum RequestType {
        SEND_IDR = 0;
        MUTE_AUDIO = 1;
        PLAY_AUDIO = 2;
    }
    RequestType type = 1;
}

message ClientDataFrameInner {
    oneof Frame {
        KeyboardFrame keyboard = 2;
        MouseFrame mouse = 3;
        ControllerFrame controller = 4;
        RequestToHost host_request = 5;
    }
}

message ClientDataFrame {
    uint32 frame_id = 1;
    bytes encrypted_data_frame = 2;
}